types.go

```diff
@@ -83,3 +64,91 @@ func (m PubKey) AsABCI() abci.PubKey {
 		Type: m.Type,
 	}
 }
+
+func (m ValidatorUpdates) Validate() error {
+	var err error
+	for _, v := range m.ValidatorUpdates {
+		err = errors.Append(err, v.Validate())
+	}
+	return err
+}
+
+// Get gets a ValidatorUpdate by a public key if it exists
+func (m ValidatorUpdates) Get(key PubKey) (ValidatorUpdate, int, bool) {
+	for k, v := range m.ValidatorUpdates {
+		if v.PubKey.Type == key.Type && bytes.Equal(v.PubKey.Data, key.Data) {
+			return v, k, true
+		}
+	}
+
+	return ValidatorUpdate{}, -1, false
+}
+
+// Deduplicate makes sure we only use the last validator update to any given validator.
+// For bookkeeping we have an option to drop validators with zero power, because those
+// are being removed by tendermint once propagated.
+func (m ValidatorUpdates) Deduplicate(dropZeroPower bool) ValidatorUpdates {
+	if len(m.ValidatorUpdates) == 0 {
+		return m
+	}
+
+	duplicates := make(map[string]int)
+	deduplicatedValidators := make([]ValidatorUpdate, 0, len(m.ValidatorUpdates))
+
+	for _, v := range m.ValidatorUpdates {
+		if key, ok := duplicates[v.PubKey.String()]; ok {
+			deduplicatedValidators[key] = v
+			continue
+		}
+		deduplicatedValidators = append(deduplicatedValidators, v)
+		duplicates[v.PubKey.String()] = len(deduplicatedValidators) - 1
+		m.ValidatorUpdates = deduplicatedValidators
+	}
+
+	if dropZeroPower {
+		filteredValidators := make([]ValidatorUpdate, 0, len(deduplicatedValidators))
+
+		for _, v := range deduplicatedValidators {
+			if v.Power != 0 {
+				filteredValidators = append(filteredValidators, v)
+			}
+		}
+		m.ValidatorUpdates = filteredValidators
+	}
+
+	return m
+}
+
+// Store stores ValidatorUpdates to the KVStore.
+func (m ValidatorUpdates) Store(store KVStore) error {
+	marshalledUpdates, err := m.Marshal()
+	if err != nil {
+		return errors.Wrap(err, "validator updates marshal")
+	}
+	err = store.Set([]byte(storeKey), marshalledUpdates)
+
+	return errors.Wrap(err, "kvstore save")
+}
+
+func GetValidatorUpdates(store KVStore) (ValidatorUpdates, error) {
+	vu := ValidatorUpdates{}
+	b, err := store.Get([]byte(storeKey))
+	if err != nil {
+		return vu, errors.Wrap(err, "kvstore get")
+	}
+
+	err = vu.Unmarshal(b)
+	return vu, errors.Wrap(err, "validator updates unmarshal")
+}
+
+func ValidatorUpdatesFromABCI(u []abci.ValidatorUpdate) ValidatorUpdates {
```

Comment:
We will still need this method, just up the chain.

Reply:
Otherwise it is used in tests only for the `abci.ResponseEndBlock` only. I don't insist on removing it but the main point here is that we can use our own type with the genesisParams to ensure a clean separation at the framework level. Anyway, PP and not a blocker.

```
type GenesisParams struct {
	Validators ValidatorUpdates
}

// FromInitChain initialises GenesisParams using abci.RequestInitChain
// data.
func FromInitChain(req abci.RequestInitChain) GenesisParams {
	return GenesisParams{
		Validators: ValidatorUpdatesFromABCI(req.Validators),
	}
}
```